### DISCUSSION TITLE
Create label for picked mouse point in an interactive manner

### DISCUSSION THREAD

**Discussion by fdinhSG** [pyvista/pyvista#4259] [Q&A]
I currently have a mesh with a scalar distribution on it. For this mesh, I'd want the min and max values shown, plus the ability to select a point and return the value on it: I want the final result to be something like the probe tool Ansys has.

I've managed to show the max and min points by saving them in an array and plotting them using `Plotter.add_point_labels`, before showing the initial plot. How can I manage to show an interactive label tool?

Ideally I'll be enabling tracking mouse position with `Plotter.enable_mouse_position()` and get the cell data using `Plotter.pick_mouse_position()`, but after that I don't know how to update the label list. I've tried setting an infinite `while True:` loop, in which at every iteration the mouse position is checked, then the label actor is destroyed and recreated using the updated labels, but it looks flickery and it lags a lot on my pc. I don't know if there's a native implementation that allows me to update the plotter content without hackish `while True:` loops.

Any help is appreciated!

(bonus question: how can I give each label a different color?)

**Comment by fdinhSG** [comment]
I managed to solve the clicking part: [enable_point_picking](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.enable_point_picking.html) has a callback as first option. I created a `update_labels` function that gets called every time I click something, and I update the labels inside that function.

I'm now wondering if this is doable with the mouse position too: is it possible to have something that updates every frame based on the position the mouse is in?

**Reply by banesullivan** [reply]
We can tie an event observer to a timer that would do the same thing. But timers are not a flushed-out API in PyVista, and the performance will be low. I recommend tieing a callback to `MouseMoveEvent`.

Further, this is going to have much better picking results after https://github.com/pyvista/pyvista/pull/4254

```py
import pyvista as pv

mesh = pv.Sphere()

pl = pv.Plotter(notebook=0)
pl.add_mesh(mesh)

# Enable mouse tracking - access `pl.mouse_position`
pl.track_mouse_position()

# Enable point picking to improve picking results
pl.enable_point_picking(show_message=False)

def on_event(*args):
    if pl.mouse_position is None:
        return
    x, y = pl.mouse_position
    pl.add_text(f'{x}, {y}', name='mouse_position_text')
    position = pl.pick_mouse_position()
    pl.add_point_labels([position], ['Mouse',], name='mouse_position_label', pickable=False)
    
# Timer callback - not a flushed out API and poor performance
# pl.iren.add_observer("TimerEvent", on_event)
# duration = 100  # milliseconds
# pl.iren.create_timer(duration, repeating=True)

# Mouse event callback, cleaner with better performance
pl.iren.add_observer("MouseMoveEvent", on_event)

pl.show()
```


https://user-images.githubusercontent.com/22067021/231490658-2dde4aa0-364d-4e69-8f89-ffd2415a171e.mov



This screenshot was taken with https://github.com/pyvista/pyvista/pull/4254 and using `enable_point_picking` to snap to points on the mesh.

There are several ways we can further improve this and clean it up, but I hope this helps get you started!

**Reply by fdinhSG** [reply]
Hi @banesullivan, thank you so much!

I would like to return the data saved on the mesh that I stored using `mesh.cell_data['my_values']`, do you know if it's feasible with the method you suggest? I saw that it's possible to return the XY coordinates of the mouse on the screen, I don't know if one could return the points on the sphere and retrieve their scalar values too...
